### ISSUE TITLE
Unimplemented `ModInt` SAWCore operators

### ISSUE THREAD

**Issue by WeeknightMVP** [GaloisInc/saw-script#1912]
While attempting to use SAW to verify the correctness of a Cryptol `property` that uses prime modular integer fields (`prime p => Z p`), SAW's translation of the property to SAWCore included the assignment `x@... = error x@... "Unimplemented: recip IntMod"`, indicating that [`IntMod` operator](https://github.com/GaloisInc/saw-script/blob/e0604caf6bf10f4aab4f21578e1fe2fa43e4d60a/cryptol-saw-core/saw/Cryptol.sawcore#L997C1-L1002C5) `recip` is currently unimplemented.

**Comment by RyanGlScott** [comment]
Indeed, this would be nice to have. Currently, all of the `IntMod` operations are implemented as SAWCore primitives, such as the ones here:

https://github.com/GaloisInc/saw-script/blob/e0604caf6bf10f4aab4f21578e1fe2fa43e4d60a/saw-core/prelude/Prelude.sawcore#L1629-L1633

At first blush, adding `recip` support for `IntMod` would mean adding a new `intModRecip` primitive and giving it appropriate semantics in each of the SAW simulator backends. This might not be entirely as straightforward to do for `recip` as the other `IntMod` primitives, since some of the backends (e.g., What4) lack direct counterparts to Cryptol's `recip` function.

There is also the question about how to translate `recip` over `IntMod`s in the Coq translation. I image that `ZModulo` in the Coq standard library would give us the tools we need to implement a version of `recip` in Coq, but I haven't looked deeply at it.

**Comment by RyanGlScott** [comment]
Apologies, my comment in https://github.com/GaloisInc/saw-script/issues/1912#issuecomment-1686069781 is somewhat misleading. Although the SAWCore prelude defines most `IntMod` operations as primitives, these primitives _do_ have semantics (see the code [here](https://github.com/GaloisInc/saw-script/blob/0f5e80afd4ea19e7451e8ab24d25463a27f9ea14/saw-core/src/Verifier/SAW/Simulator/Concrete.hs#L239-L246), [here](https://github.com/GaloisInc/saw-script/blob/0f5e80afd4ea19e7451e8ab24d25463a27f9ea14/saw-core-sbv/src/Verifier/SAW/Simulator/SBV.hs#L215-L222), and [here](https://github.com/GaloisInc/saw-script/blob/0f5e80afd4ea19e7451e8ab24d25463a27f9ea14/saw-core-what4/src/Verifier/SAW/Simulator/What4.hs#L269-L276)). The main gap right now is that the implementations of `Z`'s `Field` operations _aren't_ defined as primitives, but rather as functions that simply call `error`:

https://github.com/GaloisInc/saw-script/blob/0f5e80afd4ea19e7451e8ab24d25463a27f9ea14/cryptol-saw-core/saw/Cryptol.sawcore#L997-L1002

**Comment by RyanGlScott** [comment]
Note that these operations _are_ already implemented in Cryptol's [concrete backend](https://github.com/GaloisInc/cryptol/blob/876e97a745d157cf0048551ea7f4d2d3f1bca35a/src/Cryptol/Backend/Concrete.hs#L342-L348) and symbolic backends ([What4](https://github.com/GaloisInc/cryptol/blob/876e97a745d157cf0048551ea7f4d2d3f1bca35a/src/Cryptol/Backend/What4.hs#L659-L691) and [SBV](https://github.com/GaloisInc/cryptol/blob/876e97a745d157cf0048551ea7f4d2d3f1bca35a/src/Cryptol/Backend/SBV.hs#L420-L452)), so it should be theoretically possible to port these over to SAWCore.

One complication that arises when doing so is that Cryptol's symbolic backends implement `recip x` by generating a fresh, symbolic constant `z`, asserting the side condition that `x * z == 1`, and then returning `z`. Despite my attempts to find one, I couldn't find a straightforward way to assert side conditions directly at the SAWCore simulator level (most of the code involving side conditions requires access to a Crucible backend, which the SAWCore simulator lacks). We may need to alter the SAWCore simulator API in order to return a collection of these side conditions so that the call side of the simulator can then assert these side conditions.

**Comment by RyanGlScott** [comment]
I have opened https://github.com/GaloisInc/saw-script/issues/2433 about the task of adding side conditions during SAWCore simulation, as proposed above in https://github.com/GaloisInc/saw-script/issues/1912#issuecomment-2468893453.

**Comment by brianhuffman** [comment]
I just tried out the following definitions in `Cryptol.sawcore`:
```
intModRecip : (n : Nat) -> IntMod n -> IntMod n;
intModRecip n x =
  expByNat (IntMod n) (toIntMod n (natToInt 1)) (intModMul n) x (subNat n 2);

intModDiv : (n : Nat) -> IntMod n -> IntMod n -> IntMod n;
intModDiv n x y = intModMul n x (intModRecip n y);

PFieldIntMod : (n : Nat) -> PField (IntMod n);
PFieldIntMod n =
  { fieldRing = PRingIntMod n
  , recip     = intModRecip n
  , fieldDiv  = intModDiv n
  };
```
The idea is that for prime `n`, Fermat's Little Theorem says that `x^^(n-2)` should be a multiplicative inverse for `x : Z n`.

I can't properly evaluate how well this works, unfortunately, because SAW seems to be generally much worse than Cryptol at proving anything about modular exponentiation due to #3263.
